File: kafka/src/test/groovy/io/micronaut/configuration/kafka/docs/consumer/batch/manual/BookListener.java
package io.micronaut.configuration.kafka.docs.consumer.batch.manual;

// tag::imports[]
import io.micronaut.configuration.kafka.annotation.KafkaListener;
import io.micronaut.configuration.kafka.annotation.Topic;
import io.micronaut.configuration.kafka.docs.consumer.batch.Book;
import org.apache.kafka.clients.consumer.Consumer;
import org.apache.kafka.clients.consumer.OffsetAndMetadata;
import org.apache.kafka.common.TopicPartition;

import java.util.Collections;
import java.util.List;

import static io.micronaut.configuration.kafka.annotation.OffsetReset.EARLIEST;
import static io.micronaut.configuration.kafka.annotation.OffsetStrategy.DISABLED;
// end::imports[]

class BookListener {

    // tag::method[]
    @KafkaListener(offsetReset = EARLIEST, offsetStrategy = DISABLED, batch = true) // <1>
    @Topic("all-the-books")
    public void receive(List<Book> books,
                        List<Long> offsets,
                        List<Integer> partitions,
                        List<String> topics,
                        Consumer kafkaConsumer) { // <2>

        for (int i = 0; i < books.size(); i++) {

            // process the book
            Book book = books.get(i); // <3>

            // commit offsets
            String topic = topics.get(i);
            int partition = partitions.get(i);
            long offset = offsets.get(i); // <4>

            kafkaConsumer.commitSync(Collections.singletonMap( // <5>
                new TopicPartition(topic, partition),
                new OffsetAndMetadata(offset + 1, "my metadata")
            ));

        }
    }
    // end::method[]
}

File: test-suite-python/src/test/python/micronaut/kafka/docs/consumer/batch/manual/BookListener.py
from java.lang import Long
from java.util import Collections
from micronaut.configuration.kafka.annotation import KafkaListener, OffsetReset, OffsetStrategy, Topic
from micronaut.context.annotation import Requires
from micronaut.kafka.docs.consumer.batch.Book import Book
from org.apache.kafka.clients.consumer import Consumer, ConsumerRecord, ConsumerRecords, OffsetAndMetadata
from org.apache.kafka.common import TopicPartition


@Requires(property="spec.name", value="BatchManualAckSpec")
class BookListener:

    # tag::method[]
    @KafkaListener(offsetReset=OffsetReset.EARLIEST, offsetStrategy=OffsetStrategy.DISABLED, batch=True)  # <1>
    @Topic("all-the-books")
    def receive(self, records: list[ConsumerRecord[str, Book]], kafka_consumer: Consumer) -> None:  # <2>

        for record in records:  # <3>

            # process the book
            book = record.value()

            # commit offsets
            topic = record.topic()
            partition = record.partition()
            offset = record.offset()  # <4>

            kafka_consumer.commitSync(Collections.singletonMap(  # <5>
                TopicPartition(topic, partition),
                OffsetAndMetadata(offset + 1, "my metadata")
            ))
    # end::method[]

    # tag::consumerRecords[]
    @KafkaListener(offsetReset=OffsetReset.EARLIEST, offsetStrategy=OffsetStrategy.DISABLED, batch=True)  # <1>
    @Topic("all-the-books")
    def receive_consumer_records(self, consumer_records: ConsumerRecords[str, Book], kafka_consumer: Consumer) -> None:  # <2>
        for partition in consumer_records.partitions():  # <3>
            offset = Long.MIN_VALUE
            # process partition records
            for record in consumer_records.records(partition):  # <4>
                # process the book
                book = record.value()
                # keep last offset
                offset = record.offset()  # <5>

            # commit partition offset
            kafka_consumer.commitSync(Collections.singletonMap(  # <6>
                partition,
                OffsetAndMetadata(offset + 1, "my metadata")
            ))
    # end::consumerRecords[]
